pkg/ipam/node_manager_test.go

```diff
@@ -589,6 +590,141 @@ func TestNodeManagerAbortRelease(t *testing.T) {
 	require.Equal(t, 3, node.Stats().IPv4.UsedIPs)
 }
 
+// TestNodeManagerAbortReleaseIPReassignment tests a scenario where:
+// 1. An IP is released and marked as released in status.ipam.release-ips
+// 2. The IP is removed from spec.ipam.pool
+// 3. Before being removed from status.ipam.release-ips, the same IP is reassigned to the pool (ie. AWS ENI in the case of AWS ENI IPAM)
+// 4. The IP is no longer considered excess, so the release handshake is aborted and the IP is deleted from status.ipam.release-ips
+func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
+	var wg sync.WaitGroup
+	operatorOption.Config.ExcessIPReleaseDelay = 2
+	am := newAllocationImplementationMock()
+	require.NotNil(t, am)
+	mngr, err := NewNodeManager(hivetest.Logger(t), am, k8sapi, metricsmock.NewMockMetrics(), 10, true, false)
+	require.NoError(t, err)
+	require.NotNil(t, mngr)
+
+	// Announce node, wait for IPs to become available
+	cn := newCiliumNode("node4", 1, 3, 0)
+	mngr.Upsert(cn)
+	require.NoError(t, testutils.WaitUntil(func() bool { return reachedAddressesNeeded(mngr, "node4", 0) }, 5*time.Second))
+
+	node := mngr.Get("node4")
+	require.NotNil(t, node)
+	require.Equal(t, 3, node.Stats().IPv4.AvailableIPs)
+	require.Equal(t, 0, node.Stats().IPv4.UsedIPs)
+
+	// Use 3 out of 4 IPs
+	mngr.Upsert(updateCiliumNode(cn, 3))
+	require.NoError(t, testutils.WaitUntil(func() bool { return reachedAddressesNeeded(mngr, "node4", 0) }, 5*time.Second))
+
+	node = mngr.Get("node4")
+	require.NotNil(t, node)
+	require.Equal(t, 4, node.Stats().IPv4.AvailableIPs)
+	require.Equal(t, 3, node.Stats().IPv4.UsedIPs)
+
+	// Release one IP
+	mngr.Upsert(updateCiliumNode(node.resource, 2))
+
+	node = mngr.Get("node4")
+	require.NotNil(t, node)
+	require.Equal(t, 4, node.Stats().IPv4.AvailableIPs)
+	require.Equal(t, 2, node.Stats().IPv4.UsedIPs)
+
+	node.instanceSync.Trigger()
+
+	// This is the key IP that will be released and later reassigned
+	var releasedIP string
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		// Wait for the IP to be marked as excess for longer than the ExcessIPReleaseDelay
+		require.Eventually(t, func() bool {
+			if len(node.ipv4Alloc.ipsMarkedForRelease) == 0 {
```

Comment:
There's a race here between the test goroutine accessing `ipsMarkedForRelease` and the trigger goroutine run by the NodeManager (you can see that running the test with the race detector). It can be solved taking a read lock on the node at the start of the condition:

```go
	require.Eventually(t, func() bool {
		node.mutex.RLock()
		defer node.mutex.RUnlock()

		if len(node.ipv4Alloc.ipsMarkedForRelease) == 0 {
```